File: CSharp/LeetCode/Code209.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    public class Code209
    {
        public int MinSubArrayLen(int target, int[] nums)
        {
            int result = int.MaxValue;            
            int start = 0;
            int sum = 0;
            int length = 0;
            for (int end = 0; end < nums.Length; end++)
            {
                sum += nums[end];
                if (sum >= target)
                {
                    while(sum >= target) //开始 滑动 这是 精髓
                    {
                        length = end - start + 1;
                        result = Math.Min(result, length);

                        sum -= nums[start]; // 开始位置开始向后移动 关键
                        start += 1;
                    }
                }
            }
            return result ==int.MaxValue ? 0 : result;
        }


        [TestMethod]
        public void Test()
        {
            int target = 7;
            int[] nums = { 2, 3, 1, 2, 4, 3 };
            var result = MinSubArrayLen(target, nums);
            Console.WriteLine(result);
        }


    }
}

File: rust/src/code209.rs
pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
    let mut start = 0;    
    let mut  sum = 0;
    let mut  sub_len = 0;
    let mut  result: i32 = i32::MAX;
    for (end, value) in nums.iter().enumerate()
    {
        sum += value;
        while sum >= target {
            sub_len = end - start + 1;
            result = i32::min(result, sub_len as i32);
            sum -= nums[start as usize];
            start += 1;
        }
    }
    if result == i32::MAX{ return 0;}
    result
}
